aws/rust-runtime/aws-config/src/environment/retry_config.rs

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+use std::str::FromStr;
+
+use aws_types::os_shim_internal::Env;
+use smithy_types::retry::{RetryConfig, RetryMode};
+
+const ENV_VAR_MAX_ATTEMPTS: &str = "AWS_MAX_ATTEMPTS";
+const ENV_VAR_RETRY_MODE: &str = "AWS_RETRY_MODE";
+
+/// Load a retry_config from environment variables
+///
+/// This provider will check the values of `AWS_RETRY_MODE` and `AWS_MAX_ATTEMPTS`
+/// in order to build a retry config. If at least one is set to a valid value,
+/// construction will succeed
+#[derive(Debug, Default)]
+pub struct EnvironmentVariableRetryConfigProvider {
+    env: Env,
+}
+
+impl EnvironmentVariableRetryConfigProvider {
+    /// Create a new `EnvironmentVariableRetryConfigProvider`
+    pub fn new() -> Self {
+        EnvironmentVariableRetryConfigProvider { env: Env::real() }
+    }
+
+    #[doc(hidden)]
+    /// Create an retry_config provider from a given `Env`
+    ///
+    /// This method is used for tests that need to override environment variables.
+    pub fn new_with_env(env: Env) -> Self {
+        EnvironmentVariableRetryConfigProvider { env }
+    }
+
+    /// Attempt to create a new `RetryConfig` from environment variables
+    pub fn retry_config(&self) -> Option<RetryConfig> {
+        let max_attempts = self
+            .env
+            .get(ENV_VAR_MAX_ATTEMPTS)
+            .ok()
+            .and_then(|max_attempts| max_attempts.parse::<u32>().ok());
+        let retry_mode = self.env.get(ENV_VAR_RETRY_MODE).ok();
+
+        // If neither env vars are set, we're done with this provider
+        if let (None, None) = (&max_attempts, &retry_mode) {
+            return None;
+        }
+
+        // If at least one env var is set, we create a RetryConfig based on the value(s)
+        let mut retry_config = RetryConfig::new();
+
+        if let Some(max_attempts) = max_attempts {
+            assert_ne!(max_attempts, 0, "It is invalid to set AWS_MAX_ATTEMPTS to 0. Unset it or set it to an integer greater than or equal to one.");
+
+            retry_config = retry_config.with_max_attempts(max_attempts);
+        }
+
+        if let Some(retry_mode) = retry_mode.as_ref() {
+            match RetryMode::from_str(retry_mode).ok() {
+                Some(retry_mode) => {
+                    assert_ne!(retry_mode, RetryMode::Adaptive, r#"Setting AWS_RETRY_MODE to "adaptive" is not yet supported. Unset it or set it to a supported mode."#);
```

Comment:
this panic should probably list the supported modes. I think we might also consider just logging a warning here and ignoring the setting rather than panicing.